airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java

```diff
@@ -22,22 +26,48 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftSqlOperations.class);
   protected static final int REDSHIFT_VARCHAR_MAX_BYTE_SIZE = 65535;
 
+  private final RedshiftDataTmpTableMode redshiftDataTmpTableMode;
+  private static final List<String> tablesWithNotSuperType = new ArrayList<>();
```

Comment:
I'm not quite sure this should be kept around in such a global static list variable...

If I understand properly, it seems the issue you are facing here is that you need to detect and do a migration operation on all destination final tables that are not built with `super`  and you are relying on a discover query that returns all tables that would need to be migrated.

Unfortunately, in the `RedsfhitSqlOperations` object, the methods are called stream by the stream and you would migrate one by one, but we don't want to execute the global discover query every time for each stream?

By doing so if i am not wrong, we would start writing into tmp tables and once the sync is completed, we would start "finalizing" final tables and create/migrate them to be built with super in order to commit the data, is that correct?

1. start normal destination write in the tmp tables
2. once completed, attempt to commit into final table but check if final table exists
3. execute only once a global discovery of tables that need a migration by "caching" in a global variable of the SQLOperation object.
4. migrate them one by one while trying to commit / check if final table exist
5. commit tmp to final table

Shouldn't we move this migration logic into a proper designated pre-processing of the destination tables before starting any sync instead? So we could do:
1. global discovery of tables that need a migration
2. migrate them one by one
3. start normal destination write as usual in the tmp tables
4. once completed, commit into final table

If any migration errors occur, we would interrupt earlier and not start streaming data into tmp tables (potentially slow operation)

Reply:
@ChristopheDuong 

> If I understand properly, it seems the issue you are facing here is that you need to detect and do a migration operation on all destination final tables that are not built with super and you are relying on a discover query that returns all tables that would need to be migrated.

This is correct statement.

> Unfortunately, in the RedsfhitSqlOperations object, the methods are called stream by the stream and you would migrate one by one, but we don't want to execute the global discover query every time for each stream?

This query will be executed only once according this line of code:

```java
if (tablesWithNotSuperType.isEmpty()) {
      // we need to get the list of tables which need to be updated only once
      discoverNotSuperTables(database, schemaName);
 }
```

> we would start "finalizing" final tables and create/migrate them to be built with super in order to commit the data, is that correct?

Could you clarify what do you mean by "**finalizing**"? Now I don't have any finalizing steps at the end of sync and migration happens when we detect the tmp table with the same name and VARCHAR data type. 

In general, how my flow works right now:

1. When first stream starts it makes discovery of tables which contain VARCHAR and puts them to global list.  (BTW here I can replace this list with some Holder object)
2. After each stream checks if its name contains in that list, if so it updates column type to SUPER and migrate all the data.

Also just to be on the same page,  by "**final table**" do you mean the normalized table, right?  
Thanks.

Reply:
Hi, @ChristopheDuong!
Added to https://github.com/airbytehq/airbyte/blob/7d98dab43d5c4433fb1293b9281ad11e03410a84/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcBufferedConsumerFactory.java#L123 to `onStartFunction ` . Please, take a look.